### DISCUSSION TITLE
Questions about modeling a (lightweight) 4-bar linkage in MuJoCo

### DISCUSSION THREAD

**Discussion by miker2** [google-deepmind/mujoco#1318] [General]
I'm working on simulating the Ability Hand by Psyonic. A [github repository](https://github.com/psyonicinc/ability-hand-api/tree/master) exists with details about the hand, including the [linkage driven](https://github.com/psyonicinc/ability-hand-api/blob/master/MATLAB/get_abh_4bar_driven_angle.m) fingers.

The actual hand has 6 total actuators, despite the hand having 10 degrees of freedom. The two thumb joints are independent, but the 4 finger joints are coupled (the proximal joint determines the angle of the distal joint). Given that I have the maths required to determine the distal joint angle (q2) given the proximal joint angle (q1), I'm trying to find the best way to model this joint coupling in mujoco.

It seems overkill to try and model the individual linkages and thus the 4-bar mechanism in MuJoCo. I could do this with some equality constraints, but the numerical stability of the simulation would likely not be great.

Currently, I have all of the joint actuators for the hand set to "position" type actuators, so I have to specify each angle individually. I can think of two ways to simulate the coupling:

1. Implement the 4-bar linkage math provided in the above repository and determine the second joint angle from the first for each of the 4 fingers. Continue to use a "position" actuator for all finger joints and specify the position of the distal joint using the angle derived from the proximal joint.
2. Implement an actuator plugin. This could take a variety of forms (including what I've described in option 1) or include more complexity like deriving a joint torque for the distal joint based on the torque at the proximal joint.

I've never implemented a plugin for MuJoCo, but I've read the documentation and looked at the various examples (including the PID actuator plugin) and it seems fairly straightforward.  At this point, I'm looking for guidance on what might be the best way to proceed. I don't care about the exact coupling between the 2 joints being met; a reasonably close representation would be sufficient.

I have a more complex hand with some passive degrees of freedom that I'll be modeling next. In that case, I believe a plugin is likely the right solution, so I'm leaning towards the plugin for the Psyonic hand to gain familiarity with the implementation/integration process.

Also, if it matters I'm on version 2.6.7 of MuJoCo. I haven't upgraded to MuJoCo 3, but if that would help on the plugin front, then I can take the steps to do so.

**Comment by yuvaltassa** [comment]
Sounds like all you need is to couple joints using a [joint equality constraint](https://mujoco.readthedocs.io/en/latest/XMLreference.html#equality-joint), I don't think you need the full linkage. This would be much more stable and achieve a similar effect.

**Reply by miker2** [reply]
Thanks @yuvaltassa. I must have missed the fact that I can specify polynomial coefficients while reading the documentation. I'll give that a shot and make sure it has the desired outcome.

I do think I'll need something more complex for future work (basically there is a 4-bar with a spring in one of the links, so there is a force-dependent configuration). Would that be achievable using existing features (that don't require incorporating the links/springs in the MJCF) or would I need a plugin (or something similar) to model that coupling?

**Reply by yuvaltassa** [reply]
Yes I was just about to mention the polynomial option 🙂 

Regarding the spring in your link, it depends on how nonlinear the force is. Remember you can attach a springy spatial tendon to any two points on any two bodies, which can give you all sorts of nonlinear springy effects.

But at this point you'd probably want to do proper system ID and match data to the sim...

**Comment by jessicayin** [comment]
@miker2 I know this is an older post, but wanted to follow up to see if you can share your findings. I'm also modeling the Psyonic hand in Mujoco and would really appreciate any thoughts/tips/code that you are willing to share. Thanks!
